cli/src/action/database/upgrade.rs

```diff
@@ -0,0 +1,493 @@
+        }
+
+        let proposals = from
+            .list_proposals(predicates)
+            .map_err(ImportError::Store)?;
+        for proposal in proposals {
+            let requester = proposal.requester_node_id().to_string();
+            to.add_proposal(proposal)
+                .map_err(|e| ImportError::Proposal(requester, e))?;
+            import_result.proposals += 1;
+        }
+
+        Ok(import_result)
+    }
+
+    /// Import yaml state from the specified directory to a database
+    fn import_yaml_state_to_database(
+        state_dir: &str,
+        db_store: &'_ dyn AdminServiceStore,
+    ) -> Result<(), CliError> {
+        info!("Upgrading splinterd state");
+
+        info!("Loading YAML datastore... ");
+        fn invalid_utf8() -> CliError {
+            CliError::ActionError("'state_dir' is not a valid UTF-8 string".to_string())
+        }
+
+        let circuits_location = Path::new(&state_dir).join(CIRCUITS_FILE);
+        let proposals_location = Path::new(&state_dir).join(PROPOSALS_FILE);
+
+        if !(circuits_location.exists() || proposals_location.exists()) {
+            return Err(CliError::ActionError(
+                "no yaml state files found to import".to_string(),
+            ));
+        }
+
+        let yaml_admin_service_store = YamlAdminServiceStore::new(
+            circuits_location
+                .to_str()
+                .ok_or_else(invalid_utf8)?
+                .to_string(),
+            proposals_location
+                .to_str()
+                .ok_or_else(invalid_utf8)?
+                .to_string(),
+        )
+        .map_err(|err| {
+            CliError::ActionError(format!("unable to create YamlAdminServiceStore: {}", err))
+        })?;
+
+        info!("Processing import data... ");
+        let result = Self::import_store(db_store, &yaml_admin_service_store).map_err(|e| {
+            CliError::ActionError(match e.source() {
+                Some(source) => format!("{}: {}", e.to_string(), source),
+                None => e.to_string(),
+            })
+        })?;
+
+        info!("Backing up state files... ");
+        let new_circuits_location = circuits_location.with_extension("yaml.old");
+        let new_proposals_location = proposals_location.with_extension("yaml.old");
+        fs::rename(circuits_location, new_circuits_location)
+            .map_err(|e| CliError::ActionError(format!("could not move circuits file: {}", e)))?;
+        fs::rename(proposals_location, new_proposals_location)
+            .map_err(|e| CliError::ActionError(format!("could not move proposals file: {}", e)))?;
+
+        info!(
+            "Successfully imported {} node(s) and {} proposal(s)",
+            result.nodes, result.proposals
+        );
+
+        Ok(())
+    }
+}
+
+impl Action for ImportFromYamlAction {
+    fn run<'a>(&mut self, arg_matches: Option<&ArgMatches<'a>>) -> Result<(), CliError> {
+        let (state_dir, database_uri) = match arg_matches {
```

Comment:
If I provide a state-dir but not a database uri (connect value) I expect that database to be in the same state dir as what I provided. This does not happen now, instead it puts the database in the default location.

Reply:
It would also be useful to log the state directory being used and what database the state is being moved to for debugging purposes. (with absolute paths for files)

Reply:
Fixed!
- Code now looks for the database in the state directory if sqlite feature is enabled
- `splinter upgrade` now shows what directory it's using for the state and what database uri it's using (with absolute paths)